### ISSUE TITLE
[Messenger | ErrorHandler] error in serialization/deserialization or FlattenException::getTraceAsString incorrect return type

### ISSUE THREAD

**Issue by wucdbm** [symfony/symfony#38792]
**Symfony version(s) affected**: 5.1.7

**Description**  
<!-- A clear and concise description of the problem. -->
The issue stems from a failed Messenger message. I have a failed Doctrine transport configured. Messages, together with a serialized trace are being saved. However, when attempting to print the error via `./bin/console messenger:failed:show 6 -vv`, the command fails with an error in ErrorHandler's FlattenException class:
```
In FlattenException.php line 360:
                                                                                                                                           
  [TypeError]                                                                                                                              
  Return value of Symfony\Component\ErrorHandler\Exception\FlattenException::getTraceAsString() must be of the type string, null returned
```

I did manually decompose the data saved in the `headers` field in `messenger_messages` and it seems that `X-Message-Stamp-Symfony\Component\Messenger\Stamp\RedeliveryStamp` contains a `flattenException` field/object which only has the following fields: `type`, `title`, `status`, `detail`, `class`, `trace`.
Judging by `FlattenException::$traceAsString`'s usages, I don't see how it's empty, but I do see how it can be skipped when persisting to the database, and deserialization will afterwards not work.

Setting `getTraceAsString`'s return type to `?string` fixes that problem, but then

```
In SymfonyStyle.php line 445:
                                                                                                                                                                                                                       
  [TypeError]                                                                                                                                                                                                          
  Argument 1 passed to Symfony\Component\Console\Style\SymfonyStyle::writeBuffer() must be of the type string, null given, called in /home/wucdbm/PhpstormProjects/gems-api/vendor/symfony/console/Style/SymfonyStyle  
  .php on line 376
```

Casting `FlattenException::$traceAsString` to string works OK, but the exception trace in the command output is naturally missing.

**How to reproduce**  
<!-- Code and/or config needed to reproduce the problem. If it's a complex bug,
     create a "bug reproducer" as explained in:
     https://symfony.com/doc/current/contributing/code/reproducer.html -->

- install Symfony + messenger component
- configure `failure_transport: failed` under `framework.messenger`
- configure `failed: 'doctrine://default?queue_name=failed'` transport under `framework.messenger.transports`
- configure any other transport and route below Message to it
- create a Message
- create a handler for the Message, in my case a class implementing `__invoke()`, immediately throw an Exception and make it fail
- Update schema to create the `messenger_messages` table
- send a Message
- run `bin/console messenger:failed:show`
- run `bin/console messenger:failed:show <id> -vvv`

**Possible Solution**  
<!--- Optional: only if you have suggestions on a fix/reason for the bug -->

Possible solutions (worst to best, IMO)
1) cast $traceAsString to string (null -> '') (to my knowledge, $traceAsString can't be null in any other case but deserialization?)
2) save traceAsString in `messenger_messages` in addition to the currently persisted data
3) rebuild traceAsString from the trace array upon deserialization

I guess the correct solution to this problem would be rebuilding `$traceAsString` from the trace array as storing large text values in DBs isn't the greatest idea.
Besides, for an app that has a failed messages explorer, the trace array is sufficient to have a custom-formatted UI. I guess traceAsString only matters in CLI, so no big deal if we're rebuilding traceAsString every time we have to print it in the terminal.
Unfortunately, I have only recently started to actively use the Messenger component and I am not that familiar with its internals. On top of that, I am currently chasing strict deadlines and can't spend much time on this, at least for the foreseeable future.

**Additional context**  
<!-- Optional: any other context about the problem: log messages, screenshots, etc. -->

https://imgur.com/a/eaVMzQj

1. show messages.png
2. show message verbose.png
3. show message return type null.png
4. show message case null to empty string.png

**Comment by Nyholm** [comment]
Wow, This is an excellent issue. Thank you. 

I agree with your possible solutions. I'll try to make a PR to fix this problem for you.

**Comment by wucdbm** [comment]
Tyvm for the quick response. It's not that urgent right now, I had just discovered https://github.com/symfony/symfony/pull/35453 before looking into this, because of Doctrine errors failing any subsequent jobs.

Just that I spent a considerable amount of time tracing random behaviour in the app and I won't be able to continue debugging due to the deadlines -.-

**Comment by Nyholm** [comment]
Just a quick question. What serializer are you using?

**Comment by wucdbm** [comment]
```
framework:
    messenger:
        serializer:
            default_serializer: messenger.transport.symfony_serializer
            symfony_serializer:
                format: json
                context: { }
        failure_transport: failed

        transports:
            sync: 'sync://'

            # https://symfony.com/doc/current/messenger.html#transport-configuration
            async:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                serializer: messenger.transport.symfony_serializer
                retry_strategy:
                    max_retries: 3
                    delay: 2000
                    multiplier: 2
                    max_delay: 0
            imports:
                dsn: '%env(MESSENGER_IMPORTS_DSN)%'
                serializer: messenger.transport.symfony_serializer
                retry_strategy:
                    max_retries: 0
```

**Comment by Nyholm** [comment]
Thank you. 

I can confirm that I saw this bug in 5.1.7. 
But when Im trying to reproduce with the latest changes in 5.x then everything is working fine.

**Comment by wucdbm** [comment]
eg 5.1.8 should be fine?
In that case let's hold on until it's released and I'll test again with our current setup, otherwise I guess I'll have to prepare a dummy project where it happens

**Comment by Nyholm** [comment]
Actually no.. hm

5.2.0 will be fine. I need to dig deeper into this to make sure the fix is also in 5.1.8.

**Comment by Nyholm** [comment]
I added one PR to start with. (#38800)

The issue is that the `FlatternException` does not unserialize very well. In 5.2 dont use FlatternException on the RedeliverStamp, but on the ErrorDetailsStamp which is why I could not reproduce the error.

**Comment by wucdbm** [comment]
Awesome work, thanks again.

Whilst we're at it, is there any way to plug into Messenger's handlers and have an error subscriber (besides failed queue), in order to store exception logs somewhere else?
Maybe some sort of middleware, looking for aforementioned `ErrorDetailsStamp`?

If not, I think that'd be a nice addition in the future as it's currently very hard to debug, especially those elusive Doctrine errors.

**Comment by Jean85** [comment]
#38966 got closed in favor of #38941, which was merged into `5.x`; does this mean that it will be backported to 5.1?

**Comment by chalasr** [comment]
@Jean85 No, The `5.x` (formerly `master`) branch stands for the development branch (next 5.x version). 
See https://github.com/symfony/symfony/pull/38800#issuecomment-720115828 for alternatives to fix the bug on lower branches.

**Comment by Jean85** [comment]
Thank you @chalasr for the clarification. Is it too late to have it into 5.2?

**Comment by chalasr** [comment]
`5.x` represents 5.2 until it's released, then it'll be on its own branch. So yes, #38941 will go into 5.2.

**Comment by wucdbm** [comment]
Seems like it was fixed in #39004
Works with me on 5.2
Just a note: Already persisted messages won't work upon upgrading to 5.2. Remove those and start fresh.

Thank you @Nyholm @Jean85
